File: infrastructures/ddd/python/src/infrastructure/persistence/infra/base_entity.py
"""
Base class for all persistence entities.
Entities are database representations and CAN have ORM attributes.
They live in /infrastructure/persistence/.
"""

from abc import ABC
from datetime import datetime
from typing import Optional


class BaseEntity(ABC):
    """
    Base class for all persistence entities.
    
    Entities are responsible for:
    - Database representation
    - ORM mappings
    - Data persistence
    
    Entities CAN:
    - Have database/ORM attributes
    - Be decorated with ORM decorators
    - Reference other entities
    
    Entities MUST NOT:
    - Contain business logic
    - Be exposed outside the infrastructure layer
    """
    
    def __init__(self):
        """Initialize a new entity with default values."""
        self.id: str = ""
        self.created_at: datetime = datetime.utcnow()
        self.updated_at: datetime = datetime.utcnow()

File: infrastructures/ddd/typescript/src/infrastructure/persistence/infra/BaseEntity.ts
/**
 * Base class for all persistence entities.
 * This class:
 * - Lives in /infrastructure/persistence/infra/
 * - Contains common database fields
 * - CAN have database/ORM attributes
 * - Is extended by BaseWriteEntity and BaseReadEntity
 * 
 * Entities are persistence models that:
 * - Map to database tables/collections
 * - Have ORM/database attributes
 * - Are separate from business models (BMOs)
 * - Are transformed via mappers
 */
export abstract class BaseEntity {
  public Id: string;
  public CreatedAt: Date;
  public UpdatedAt: Date;

  constructor() {
    this.Id = '';
    this.CreatedAt = new Date();
    this.UpdatedAt = new Date();
  }
}
